compiler/rustc_hir_analysis/src/collect/type_of.rs

```diff
@@ -786,15 +792,24 @@ fn find_opaque_ty_constraints_for_rpit(
             // the `concrete_opaque_types` table.
             tcx.ty_error(guar)
         } else {
-            table.concrete_opaque_types.get(&def_id).map(|ty| ty.ty).unwrap_or_else(|| {
-                // We failed to resolve the opaque type or it
-                // resolves to itself. We interpret this as the
-                // no values of the hidden type ever being constructed,
-                // so we can just make the hidden type be `!`.
-                // For backwards compatibility reasons, we fall back to
-                // `()` until we the diverging default is changed.
-                tcx.mk_diverging_default()
-            })
+            table
+                .concrete_opaque_types
+                .iter()
+                .find(|(key, _)| {
+                    // For backwards compatibility, we choose the first matching
+                    // opaque type definition
```

Comment:
What happens if we check all of them?

Reply:
It's not that we have to check them, it's that we have to merge them. I guess I could do that.

Reply:
r=me with that

Reply:
Wait wait wait... We've remapped all of the hidden types back to their declaration params. We don't need to put the substs in the writeback results at all. :facepalm: